config/script.go

```diff
@@ -99,102 +145,140 @@ func evalScriptDesc(evalctx *eval.Context, expr hhcl.Expression, name string) (s
 	}
 
-	if !val.Type().IsTupleType() {
-		return nil, errors.E(ErrScriptInvalidTypeCommand, expr.Range(),
-			"%s should be a list(string) type", name)
-	}
-
-	if val.LengthInt() == 0 {
+	if cmdList.LengthInt() == 0 {
 		return nil, errors.E(ErrScriptEmptyCmds, expr.Range())
 	}
 
 	errs := errors.L()
-	evaluatedCommand := []string{}
+	evaluatedCommands := []*ScriptCmd{}
 
 	index := -1
-	it := val.ElementIterator()
+	it := cmdList.ElementIterator()
 	for it.Next() {
 		index++
 		_, elem := it.Element()
-		if elem.Type() != cty.String {
-			errs.Append(errors.E(ErrScriptInvalidTypeCommand, expr.Range(),
-				"field %s must be a list(string) but element %d has type %q",
-				name, index, elem.Type().FriendlyName()))
+		if !elem.Type().IsTupleType() {
+			errs.Append(errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
+				"commands must be a list of list, but element %d has type %q",
+				index, elem.Type().FriendlyName()))
 			continue
 		}
-		evaluatedCommand = append(evaluatedCommand, elem.AsString())
+
+		if elem.LengthInt() == 0 {
+			errs.Append(errors.E(ErrScriptEmptyCmds, expr.Range(), "commands item %d is empty", index))
+			continue
+		}
+
+		evaluatedCommand, err := unmarshalScriptJobCommand(elem, expr)
+		if err != nil {
+			errs.Append(err)
+			continue
+		}
+
+		evaluatedCommands = append(evaluatedCommands, evaluatedCommand)
 	}
 
 	if err := errs.AsError(); err != nil {
 		return nil, err
 	}
 
-	return evaluatedCommand, nil
+	return evaluatedCommands, nil
 }
 
-func evalScriptJobCommands(evalctx *eval.Context, expr hhcl.Expression, name string) ([][]string, error) {
-	val, err := evalctx.Eval(expr)
-	if err != nil {
-		return nil, errors.E(ErrScriptSchema, expr.Range(),
-			err, "evaluating %s", name)
-	}
-
-	if !val.Type().IsTupleType() {
-		return nil, errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
-			"%s should be a list(string) type", name)
+func unmarshalScriptJobCommand(cmdValues cty.Value, expr hhcl.Expression) (*ScriptCmd, error) {
+	if !cmdValues.Type().IsTupleType() {
+		return nil, errors.E(ErrScriptInvalidTypeCommand, expr.Range(), "command must be a list")
 	}
 
-	if val.LengthInt() == 0 {
+	if cmdValues.LengthInt() == 0 {
 		return nil, errors.E(ErrScriptEmptyCmds, expr.Range())
 	}
 
 	errs := errors.L()
-	evaluatedCommands := [][]string{}
+	r := &ScriptCmd{}
 
-	index := -1
-	it := val.ElementIterator()
+	index := 0
+	lastIndex := cmdValues.LengthInt() - 1
+
+	it := cmdValues.ElementIterator()
 	for it.Next() {
-		index++
 		_, elem := it.Element()
-		if !elem.Type().IsTupleType() {
-			errs.Append(errors.E(ErrScriptInvalidTypeCommands, expr.Range(),
-				"field %s must be a list of list(string) but element %d has type %q",
-				name, index, elem.Type().FriendlyName()))
-			continue
+		if elem.Type() == cty.String {
+			r.Args = append(r.Args, elem.AsString())
+		} else if index == lastIndex {
+			if elem.Type().IsObjectType() {
```

Comment:
Considering a scenario where the options object is provided as a first element (or any position other than the last position), the resulting error will be the one on L224. 
We could catch that case with something like:

```suggestion
		} else if elem.Type().IsObjectType() {
		        if index != lastIndex {
		          errs.Append(errors.E(ErrScriptInvalidOptionsPosition, ..)
		          continue
		        }
```

Reply:
I decided not to give any specific error for this, because I think it would be very unlikely to accidentally put the option anywhere but at the end, i.e. something like this would not happen:
```
[
  ["echo", "hello", {
      cloud_sync_deployment = "false"
    }, "world"],
]
```
What's more likely is that maybe a function is used to return one of the arguments, and this function returns the wrong type, in which case the "must be a list of strings" error is better than "options must be last element", which could be really confusing in particular if you didn't even use any options, which will probably be true for most of the script commands.

Reply:
Makes sense, thanks for the clarification.